### ISSUE TITLE
Suspend to disk does not work with zfs/spl 0.7.5

### ISSUE THREAD

**Issue by jhyeon** [openzfs/zfs#7071]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Funtoo
Distribution Version    | 11 Jan 2018
Linux Kernel                 | 4.14.14
Architecture                 | Kaby Lake
ZFS Version                  | 0.7.5 from repo
SPL Version                  | 0.7.5 from repo

### Describe the problem you're observing

After upgrading zfs/spl from 0.6.5.11 to 0.7.5, userland software suspend to disk stopped working.  The messages from `s2disk` on the console can be
```
Looking for splash system... none
s2disk: Snapshotting system
s2disk: System snapshot ready. Preparing to write
s2disk: Image size: 2192764 kilobytes
s2disk: Free swap: 64611752 kilobytes
s2disk: Saving 548191 image data pages (press backspace to abort) ... 31%
```
The progress in the last line increases up to 30% or so where `s2disk` fails to hibernate and returns.  A similar thing could happen with zfs/spl 0.6.5.11 (+ kernel 4.14.14) but `s2disk` completed hibernation if I repeated it a few more times.  With zfs/spl 0.7.5, the progress always stops around 30% even if I run `s2disk` multiple times in a row.  `echo 3 > /proc/sys/vm/drop_caches` does not help.

### Describe how to reproduce the problem

1. Fill the RAM to some extent by running several applications
1. Run `s2disk`

If the RAM usage is very low e.g. just after booting the system, `s2disk` can happen to succeed.  The following is the output of `free` after an `s2disk` failure:
```
              total        used        free      shared  buff/cache   available
Mem:       32727300     1276112    31395968       30428       55220    31183392
Swap:      67108860     2678104    64430756
```
My `/etc/suspend.conf ` reads
```
snapshot device = /dev/snapshot
resume device = /dev/disk/by-label/swap
compress = y
early writeout = y
splash = y
threads = y
```

### Include any warning/errors/backtraces from the system logs

Part of `dmesg` output even though it contains no warning:
```
[  321.958683] Syncing filesystems ... 
[  327.083264] done.
[  327.083266] Freezing user space processes ... (elapsed 0.001 seconds) done.
[  327.085016] OOM killer disabled.
[  327.085154] PM: Marking nosave pages: [mem 0x00000000-0x00000fff]
[  327.085154] PM: Marking nosave pages: [mem 0x00058000-0x00058fff]
[  327.085155] PM: Marking nosave pages: [mem 0x0008c000-0x000fffff]
[  327.085156] PM: Marking nosave pages: [mem 0x65033000-0x65034fff]
[  327.085157] PM: Marking nosave pages: [mem 0x6e972000-0x6fffefff]
[  327.085211] PM: Marking nosave pages: [mem 0x70000000-0xffffffff]
[  327.086056] PM: Basic memory bitmaps created
[  327.086129] PM: Preallocating image memory... done (allocated 569280 pages)
[  329.441452] PM: Allocated 2277120 kbytes in 2.35 seconds (968.98 MB/s)
[  329.441453] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[  329.443328] wlan0: deauthenticating from a0:63:91:b1:fc:e0 by local choice (Reason: 3=DEAUTH_LEAVING)
[  329.916974] ACPI: Preparing to enter system sleep state S4
[  330.143422] ACPI: EC: event blocked
[  330.143423] ACPI: EC: EC stopped
[  330.143423] PM: Saving platform NVS memory
[  330.143976] Disabling non-boot CPUs ...
[  330.171136] smpboot: CPU 1 is now offline
[  330.194461] smpboot: CPU 2 is now offline
[  330.227780] smpboot: CPU 3 is now offline
[  330.229875] PM: Creating hibernation image:
[  330.541140] PM: Need to copy 547121 pages
[  330.541141] PM: Normal pages needed: 547121 + 1024, available pages: 7779736
[  331.100595] PM: Hibernation image created (547121 pages copied)
[  330.230591] PM: Restoring platform NVS memory
[  330.230786] ACPI: EC: EC started
[  330.231760] Enabling non-boot CPUs ...
[  330.231779] x86: Booting SMP configuration:
[  330.231779] smpboot: Booting Node 0 Processor 1 APIC 0x2
[  330.232340]  cache: parent cpu1 should not be sleeping
[  330.232416] CPU1 is up
[  330.232424] smpboot: Booting Node 0 Processor 2 APIC 0x1
[  330.233076]  cache: parent cpu2 should not be sleeping
[  330.233156] CPU2 is up
[  330.233165] smpboot: Booting Node 0 Processor 3 APIC 0x3
[  330.233757]  cache: parent cpu3 should not be sleeping
[  330.233827] CPU3 is up
[  330.236749] ACPI: Waking up from system sleep state S4
```

**Comment by bunder2015** [comment]
is `/dev/disk/by-label/swap` a zvol?  i was under the impression that hibernate never worked if swap was a zvol.  a little OT, but which initrd/init combo are you using?

**Comment by jhyeon** [comment]
It is a GUID partition on an SSD.  I have never tried hibernating to a zvol.  I am using dracut/openrc.

**Comment by bunder2015** [comment]
thanks for the info.

did you change anything other than the zfs version?  if you're not suspending to a zvol, i'm not entirely sure how zfs could be involved.  you might want to try getting an strace out of s2disk, but that might be a little hard since you're in the middle of shutting the system down.

**Comment by jhyeon** [comment]
Indeed, when I did `strace -o s2disk.txt s2disk` the system froze after printing
```
Looking for splash system... none
```

I really changed only the zfs version.  I applied patches to spl and zfs to avoid build failure when building zfs 0.6.5.11 against kernel 4.14.14.  But hibernation worked even without the patches with older kernels.

My guess as an end-user is that this might have something to do with the (new) way zfs 0.7.x manages the memory.
